test/e2e/persistent_volumes.go

```diff
@@ -41,110 +71,150 @@ func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
 	if config.serverImage != "" {
 		podName := config.prefix + "-server"
 		err := podClient.Delete(podName, nil)
-		if err != nil {
-			framework.Logf("Delete of %v pod failed: %v", podName, err)
-		}
+		Expect(err).NotTo(HaveOccurred())
 	}
 }
 
-// Delete the PV. Fail test if delete fails. If success the returned PV should
-// be nil, which prevents the AfterEach from attempting to delete it.
-func deletePersistentVolume(c *client.Client, pv *api.PersistentVolume) (*api.PersistentVolume, error) {
+// Cleanup up pvs and pvcs in multi-pv-pvc test cases. All entries found in the pv and
+// claims maps are deleted.
+// Note: this is the only code that deletes PV objects.
+func pvPvcCleanup(c *client.Client, ns string, pvols pvmap, claims pvcmap) {
+
+	if c != nil && len(ns) > 0 {
+		framework.Logf("AfterTest: deleting %v PVCs from map", len(claims))
+		for pvcKey := range claims {
+			delete(claims, pvcKey) // ok to delete key from map now
+			Expect(len(pvcKey)).NotTo(BeZero())
+			nSpace := pvcKey.NS()
+			name := pvcKey.Name()
+			_, err := c.PersistentVolumeClaims(nSpace).Get(name)
+			if !apierrs.IsNotFound(err) {
+				Expect(err).NotTo(HaveOccurred())
```

Comment:
@saad-ali I tested a failure in an AfterEach. In the log you'll see "Failure in Spec Teardown (AfterEach) ..." to indicate that the test failed inside an AfterEach. A failure in an AfterEach exits the AfterEach block and reports the failure. There can be a stack of AfterEach()s and thus a higher-level AfterEach may be invoked. For example, if a Context() AfterEach fails then the KubeDescribe AfterEach may be invoked. The final AfterEach is defined in the framework. This AfterEach deletes the namespace and this deletes the PVCs in that namespace. So, as long as all of the test claims live in the same ns they will be deleted, at least when the ns is deleted. However, our plan is to create pvcs and pod in several different namespaces, and in that case the claims residing in the other namespaces, other than the ns where the initial ns, will survive the test.  In fact, all the of the pvs (in the system global ns, remain after the test fails.